ludwig/models/llm.py

```diff
@@ -268,10 +308,37 @@ def forward_train(self, input_ids, target_ids) -> Dict[str, torch.Tensor]:
         # Get predictions, probabilities and logits tensor from the output feature's predictions function
         outputs = self.output_features.get(of_name).predictions(outputs, of_name)
 
+        # Cast to float32 for metric computation incase we're using deespeed with
+        # reduced precision such as bfloat16.
+        for prediction_key, prediction_tensor in outputs.items():
+            outputs[prediction_key] = prediction_tensor.type(torch.float32)
+
         return outputs
 
-    def forward_generate(self, input_ids, mask) -> Dict[str, torch.Tensor]:
+    def generate(
+        self,
+        inputs: Union[
+            Dict[str, torch.Tensor], Dict[str, np.ndarray], Tuple[Dict[str, torch.Tensor], Dict[str, torch.Tensor]]
+        ],
+        mask=None,
+    ) -> Dict[str, torch.Tensor]:
         """Generates tokens using the model."""
+
+        if isinstance(inputs, tuple):
+            inputs, targets = inputs
+            # Convert targets to tensors.
+            for target_feature_name, target_value in targets.items():
+                if not isinstance(target_value, torch.Tensor):
+                    targets[target_feature_name] = torch.from_numpy(target_value)
+                else:
+                    targets[target_feature_name] = target_value
+        else:
+            targets = None
+
+        assert list(inputs.keys()) == self.input_features.keys()
```

Comment:
nit: this can probably move into it's own utility function that gets shared between `generate` and `forward`, but we can do that in another PR